docs/_docs/02_testing-best-practices/unit-integration-end-to-end-test.md

```diff
@@ -22,3 +22,149 @@ Link to the video at [infoq.com](https://www.infoq.com/presentations/automated-t
+  Validators []Validator
+}
+
+func UnitTest(t *testing.T, testPlan *terraform.UnitTestPlan)
+```
+
+Examples:
+
+This example shows the simplest example with multiple Validators, with the default working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestPlan{
+    TerraformOptions: terraformOptions,
+    Validators: []terraform.Validator{
+      // A list of validation functions to test your module
+      validateWebServer,
+      valudateImageHosting,
+      validateSSLApplied,
+    },
+  })
+}
+
+func validateWebServer(t *testing.T, outputs *terraform.ModuleOutputs, workingDir string) {
+}
+```
+
+This example shows using Setup and Validators, with a custom working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestPlan{
+    TerraformOptions: terraformOptions,
+    WorkingDirectory "../web-server",
+    Setup: setup,
+    Validators: []terraform.Validator{
+      // A list of validation functions to test your module
+      validateWebServer,
+      valudateImageHosting,
+      validateSSLApplied,
+    },
+  })
+}
+
+func validateWebServer(t *testing.T, outputs *terraform.ModuleOutputs, workingDir string) {
+  ...
+}
+
+func setup(t *testing.T, workingDir string) {
+  ...
+}
+```
+
+
+## Integration test helpers
+
+**_TODO: As this is a one day project I won't attempt to complete this today _**
+
+Integration tests are very similar to unit tests, only you are testing multiple modules together. To facilitate this
+terraform provides an IntegrationTest function, the main difference to the UnitTest is the IntegrationTestPlan takes a
+list of modules, each with a unique name. It passes a map of ModuleOutputs which are keyed on the unique name you gave
+the module.
+
+The destroys will be performed in the reverse order to applies.
+
+The working directory will contain sub directories for each module (named by your unique name) to save the
+terraformOptions and outputs to.
+
+If you skip the init_apply step then IntegrationTest will automatically load your previously applied outputs from the working directory.
+
+Signature of IntegrationTest:
+
+```go
+// Output of every module applied, by name
+type AllModuleOutputs map[string]*ModuleOutputs
+
+type ModuleUnderTest struct {
+  Name  string
+  TerraformOptions *terraform.Options
+}
+
+IntegrationTest(t *testing.T, testPlan *terraform.IntegrationTestPlan
+
+// A validation function to test your terraform code
+type IntegrationTestValidator func(t *testing.T, allModuleOutputs *terraform.AllModuleOutputs, workingDir string)
+
+type IntegrationTestPlan struct {
+  Modules []*ModuleUnderTest

```

Comment:
Hmm, I was thinking a bit more about this, often you want outputs from one module to setup another module, so maybe on integration tests there should be callback functions to generate the terraform options, they could take the outputs of all previously applied modules, so like:

```
ModuleUnderTest struct {
    Name string
    OptionsGenerator func(t *testing.T, allModuleOutputs *terraform.AllModuleOutputs, workingDir string) *terraform.Options
}
```

Reply:
Was just about to bring this very thing up!